File: LibreChat/api/server/middleware/supabaseAuth.js
// Supabase JWT Verification Middleware for LibreChat Backend
const { createClient } = require('@supabase/supabase-js');
const jwt = require('jsonwebtoken');

// Initialize Supabase client
const SUPABASE_URL = process.env.SUPABASE_URL || 'https://qctdaaezghvqnbpghinr.supabase.co';
const SUPABASE_SERVICE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY || '';
const SUPABASE_JWT_SECRET = process.env.SUPABASE_JWT_SECRET || '';

const supabase = createClient(SUPABASE_URL, SUPABASE_SERVICE_KEY, {
  auth: {
    autoRefreshToken: false,
    persistSession: false,
  },
});

/**
 * Middleware to verify Supabase JWT tokens
 * Integrates with existing LibreChat auth system
 */
async function verifySupabaseToken(req, res, next) {
  try {
    // Get token from header
    const authHeader = req.headers.authorization;
    if (!authHeader || !authHeader.startsWith('Bearer ')) {
      // Fallback to existing auth method
      return next();
    }

    const token = authHeader.substring(7);

    // Try to verify as Supabase token first
    if (token.includes('eyJ') && token.split('.').length === 3) {
      try {
        // Verify JWT signature
        const decoded = jwt.decode(token, { complete: true });

        if (decoded && decoded.payload.iss === 'https://qctdaaezghvqnbpghinr.supabase.co/auth/v1') {
          // This is a Supabase token
          const { data: { user }, error } = await supabase.auth.getUser(token);

          if (error || !user) {
            return res.status(401).json({ error: 'Invalid Supabase token' });
          }

          // Get user profile from Supabase
          const { data: profile, error: profileError } = await supabase
            .from('profiles')
            .select(`
              *,
              organizations (
                id,
                name,
                slug,
                settings
              )
            `)
            .eq('id', user.id)
            .single();

          if (profileError || !profile) {
            return res.status(401).json({ error: 'User profile not found' });
          }

          // Convert Supabase user to LibreChat user format
          req.user = {
            id: user.id,
            email: user.email,
            username: profile.username || user.email.split('@')[0],
            name: profile.full_name || user.email,
            avatar: profile.avatar_url,
            role: convertSupabaseRole(profile.role),
            provider: 'supabase',
            organizationId: profile.organization_id,
            organizationName: profile.organizations?.name,
            supabaseProfile: profile,
          };

          // Add organization context
          req.organizationId = profile.organization_id;
          req.organizationRole = profile.role;

          // Check if user exists in LibreChat database
          const User = require('~/models/User');
          let libreChatUser = await User.findOne({ email: user.email });

          if (!libreChatUser) {
            // Create user in LibreChat database
            libreChatUser = await User.create({
              email: user.email,
              username: profile.username || user.email.split('@')[0],
              name: profile.full_name || user.email,
              avatar: profile.avatar_url,
              provider: 'supabase',
              supabaseId: user.id,
              organizationId: profile.organization_id,
            });
          } else {
            // Update user info
            libreChatUser.organizationId = profile.organization_id;
            libreChatUser.supabaseId = user.id;
            await libreChatUser.save();
          }

          req.user.mongoId = libreChatUser._id;
          return next();
        }
      } catch (err) {
        console.error('Supabase token verification error:', err);
        // Not a Supabase token, continue with normal auth
      }
    }

    // Continue with existing LibreChat auth
    next();
  } catch (error) {
    console.error('Auth middleware error:', error);
    next();
  }
}

/**
 * Middleware to check organization permissions
 */
function requireOrganization(req, res, next) {
  if (!req.organizationId) {
    return res.status(403).json({ error: 'Organization context required' });
  }
  next();
}

/**
 * Middleware to check admin role
 */
function requireOrgAdmin(req, res, next) {
  if (!req.organizationRole || !['org_admin', 'super_admin'].includes(req.organizationRole)) {
    return res.status(403).json({ error: 'Admin privileges required' });
  }
  next();
}

/**
 * Middleware to check resource permissions
 */
async function checkResourcePermission(resourceType, permissionLevel) {
  return async (req, res, next) => {
    try {
      const resourceId = req.params.id || req.body.resourceId;
      const userId = req.user?.id;
      const organizationId = req.organizationId;

      if (!resourceId || !userId) {
        return res.status(403).json({ error: 'Access denied' });
      }

      // Check if user has permission for this resource
      const { data: permission, error } = await supabase
        .from('resource_permissions')
        .select('*')
        .eq('resource_id', resourceId)
        .eq('resource_type', resourceType)
        .eq('organization_id', organizationId)
        .or(`grantee_id.eq.${userId},grantee_id.eq.${req.user?.supabaseProfile?.organizational_unit_id}`)
        .single();

      if (error || !permission) {
        // Check if user is resource owner
        const isOwner = await checkResourceOwnership(resourceId, resourceType, userId);
        if (!isOwner) {
          return res.status(403).json({ error: 'Permission denied' });
        }
      }

      // Check permission level
      if (permission && !hasRequiredPermission(permission.permission_level, permissionLevel)) {
        return res.status(403).json({ error: 'Insufficient permissions' });
      }

      next();
    } catch (error) {
      console.error('Permission check error:', error);
      res.status(500).json({ error: 'Permission check failed' });
    }
  };
}

/**
 * Helper function to check resource ownership
 */
async function checkResourceOwnership(resourceId, resourceType, userId) {
  // This would check the actual resource tables
  // For now, returning true for demo
  return true;
}

/**
 * Helper function to check permission levels
 */
function hasRequiredPermission(userLevel, requiredLevel) {
  const levels = {
    viewer: 1,
    editor: 2,
    admin: 3,
  };

  return levels[userLevel] >= levels[requiredLevel];
}

/**
 * Convert Supabase role to LibreChat role
 */
function convertSupabaseRole(supabaseRole) {
  const roleMap = {
    super_admin: 'ADMIN',
    org_admin: 'USER',
    member: 'USER',
  };
  return roleMap[supabaseRole] || 'USER';
}

/**
 * Sync user data between Supabase and LibreChat
 */
async function syncUserData(req, res, next) {
  try {
    if (!req.user?.provider === 'supabase') {
      return next();
    }

    // Sync any changes from Supabase to LibreChat
    const User = require('~/models/User');
    const libreChatUser = await User.findById(req.user.mongoId);

    if (libreChatUser && req.user.supabaseProfile) {
      // Update user data if changed
      let updated = false;

      if (libreChatUser.name !== req.user.supabaseProfile.full_name) {
        libreChatUser.name = req.user.supabaseProfile.full_name;
        updated = true;
      }

      if (libreChatUser.avatar !== req.user.supabaseProfile.avatar_url) {
        libreChatUser.avatar = req.user.supabaseProfile.avatar_url;
        updated = true;
      }

      if (updated) {
        await libreChatUser.save();
      }
    }

    next();
  } catch (error) {
    console.error('User sync error:', error);
    next();
  }
}

module.exports = {
  verifySupabaseToken,
  requireOrganization,
  requireOrgAdmin,
  checkResourcePermission,
  syncUserData,
};

File: memory-agent/app/middleware/supabase_auth.py
"""
Supabase JWT Authentication Middleware for Memory Agent
"""
from typing import Optional, Dict, Any
from fastapi import Request, HTTPException, Depends
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
from supabase import create_client, Client
from functools import lru_cache
import jwt
import os
from datetime import datetime
import logging

logger = logging.getLogger(__name__)

# Supabase configuration
SUPABASE_URL = os.getenv("SUPABASE_URL", "https://qctdaaezghvqnbpghinr.supabase.co")
SUPABASE_ANON_KEY = os.getenv("SUPABASE_ANON_KEY", "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6InFjdGRhYWV6Z2h2cW5icGdoaW5yIiwicm9sZSI6ImFub24iLCJpYXQiOjE3NTgyNzU3NTMsImV4cCI6MjA3Mzg1MTc1M30.WJVWs7aruIuo_jpa6o0xXbXefN8DCIK_1CTr_afVRos")
SUPABASE_SERVICE_KEY = os.getenv("SUPABASE_SERVICE_ROLE_KEY", "")
SUPABASE_JWT_SECRET = os.getenv("SUPABASE_JWT_SECRET", "")

# Initialize Supabase client
@lru_cache()
def get_supabase_client() -> Client:
    """Get cached Supabase client instance."""
    return create_client(
        SUPABASE_URL,
        SUPABASE_SERVICE_KEY or SUPABASE_ANON_KEY,
        options={
            "auto_refresh_token": False,
            "persist_session": False,
        }
    )

# HTTP Bearer for token extraction
security = HTTPBearer()

class AuthContext:
    """Authentication context containing user and organization information."""

    def __init__(self, user_data: Dict[str, Any], profile_data: Dict[str, Any]):
        self.user_id = user_data.get("id")
        self.email = user_data.get("email")
        self.user_data = user_data

        self.profile = profile_data
        self.organization_id = profile_data.get("organization_id")
        self.organizational_unit_id = profile_data.get("organizational_unit_id")
        self.role = profile_data.get("role", "member")
        self.username = profile_data.get("username") or self.email
        self.full_name = profile_data.get("full_name")

        # Organization details
        self.organization = profile_data.get("organizations", {})
        self.organization_name = self.organization.get("name")
        self.organization_slug = self.organization.get("slug")

    def is_admin(self) -> bool:
        """Check if user is an admin (org_admin or super_admin)."""
        return self.role in ["org_admin", "super_admin"]

    def is_super_admin(self) -> bool:
        """Check if user is a super admin."""
        return self.role == "super_admin"

    def to_dict(self) -> Dict[str, Any]:
        """Convert context to dictionary."""
        return {
            "user_id": self.user_id,
            "email": self.email,
            "username": self.username,
            "full_name": self.full_name,
            "role": self.role,
            "organization_id": self.organization_id,
            "organization_name": self.organization_name,
            "organizational_unit_id": self.organizational_unit_id,
        }

async def verify_supabase_token(
    credentials: HTTPAuthorizationCredentials = Depends(security)
) -> AuthContext:
    """
    Verify Supabase JWT token and return authentication context.

    Args:
        credentials: HTTP Bearer token from request

    Returns:
        AuthContext with user and organization information

    Raises:
        HTTPException: If token is invalid or user not found
    """
    token = credentials.credentials

    try:
        # Get Supabase client
        supabase = get_supabase_client()

        # Verify token with Supabase
        user_response = supabase.auth.get_user(token)

        if not user_response or not user_response.user:
            raise HTTPException(
                status_code=401,
                detail="Invalid authentication token"
            )

        user = user_response.user

        # Get user profile with organization data
        profile_response = supabase.table("profiles").select(
            "*, organizations(id, name, slug, settings), organizational_units(id, name)"
        ).eq("id", user.id).single().execute()

        if not profile_response.data:
            raise HTTPException(
                status_code=401,
                detail="User profile not found"
            )

        profile = profile_response.data

        # Create and return auth context
        return AuthContext(user.dict(), profile)

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Token verification error: {str(e)}")
        raise HTTPException(
            status_code=401,
            detail="Authentication failed"
        )

async def get_current_user(
    auth_context: AuthContext = Depends(verify_supabase_token)
) -> Dict[str, Any]:
    """
    Get current authenticated user.

    Args:
        auth_context: Authentication context from token verification

    Returns:
        User information dictionary
    """
    return auth_context.to_dict()

async def require_organization(
    auth_context: AuthContext = Depends(verify_supabase_token)
) -> AuthContext:
    """
    Require that user belongs to an organization.

    Args:
        auth_context: Authentication context

    Returns:
        AuthContext if user has organization

    Raises:
        HTTPException: If user doesn't belong to organization
    """
    if not auth_context.organization_id:
        raise HTTPException(
            status_code=403,
            detail="Organization membership required"
        )

    return auth_context

async def require_admin(
    auth_context: AuthContext = Depends(verify_supabase_token)
) -> AuthContext:
    """
    Require admin privileges (org_admin or super_admin).

    Args:
        auth_context: Authentication context

    Returns:
        AuthContext if user is admin

    Raises:
        HTTPException: If user is not admin
    """
    if not auth_context.is_admin():
        raise HTTPException(
            status_code=403,
            detail="Admin privileges required"
        )

    return auth_context

async def require_super_admin(
    auth_context: AuthContext = Depends(verify_supabase_token)
) -> AuthContext:
    """
    Require super admin privileges.

    Args:
        auth_context: Authentication context

    Returns:
        AuthContext if user is super admin

    Raises:
        HTTPException: If user is not super admin
    """
    if not auth_context.is_super_admin():
        raise HTTPException(
            status_code=403,
            detail="Super admin privileges required"
        )

    return auth_context

class ACLChecker:
    """Access Control List checker for resource permissions."""

    def __init__(self, resource_type: str, permission_level: str = "viewer"):
        self.resource_type = resource_type
        self.permission_level = permission_level

    async def __call__(
        self,
        request: Request,
        auth_context: AuthContext = Depends(verify_supabase_token)
    ) -> bool:
        """
        Check if user has required permission for resource.

        Args:
            request: FastAPI request object
            auth_context: Authentication context

        Returns:
            True if permission granted

        Raises:
            HTTPException: If permission denied
        """
        # Extract resource ID from request
        resource_id = None

        # Try to get from path parameters
        if hasattr(request, "path_params"):
            resource_id = request.path_params.get("resource_id") or \
                         request.path_params.get("id")

        # Try to get from query parameters
        if not resource_id and hasattr(request, "query_params"):
            resource_id = request.query_params.get("resource_id")

        # Try to get from body (for POST/PUT requests)
        if not resource_id and request.method in ["POST", "PUT"]:
            try:
                body = await request.json()
                resource_id = body.get("resource_id")
            except:
                pass

        if not resource_id:
            raise HTTPException(
                status_code=400,
                detail="Resource ID not provided"
            )

        # Check permissions in database
        supabase = get_supabase_client()

        # Check direct user permission
        user_perm = supabase.table("resource_permissions").select("*").eq(
            "resource_id", resource_id
        ).eq(
            "resource_type", self.resource_type
        ).eq(
            "grantee_id", auth_context.user_id
        ).eq(
            "grantee_type", "user"
        ).execute()

        if user_perm.data:
            permission = user_perm.data[0]
            if self._has_required_permission(permission["permission_level"]):
                return True

        # Check organizational unit permission
        if auth_context.organizational_unit_id:
            unit_perm = supabase.table("resource_permissions").select("*").eq(
                "resource_id", resource_id
            ).eq(
                "resource_type", self.resource_type
            ).eq(
                "grantee_id", auth_context.organizational_unit_id
            ).eq(
                "grantee_type", "organizational_unit"
            ).execute()

            if unit_perm.data:
                permission = unit_perm.data[0]
                if self._has_required_permission(permission["permission_level"]):
                    return True

        # Check if user is admin (admins have access to all resources in org)
        if auth_context.is_admin():
            return True

        # Permission denied
        raise HTTPException(
            status_code=403,
            detail=f"Insufficient permissions for {self.resource_type}"
        )

    def _has_required_permission(self, user_level: str) -> bool:
        """Check if user permission level meets requirements."""
        levels = {
            "viewer": 1,
            "editor": 2,
            "admin": 3
        }

        return levels.get(user_level, 0) >= levels.get(self.permission_level, 1)

# Convenience dependency functions
def require_memory_viewer():
    """Require viewer access to memory resource."""
    return ACLChecker("memory", "viewer")

def require_memory_editor():
    """Require editor access to memory resource."""
    return ACLChecker("memory", "editor")

def require_memory_admin():
    """Require admin access to memory resource."""
    return ACLChecker("memory", "admin")

# Metadata filter for LlamaIndex queries
def get_organization_filter(
    auth_context: AuthContext = Depends(require_organization)
) -> Dict[str, Any]:
    """
    Get metadata filter for organization-scoped queries.

    Args:
        auth_context: Authentication context

    Returns:
        Metadata filter dictionary for LlamaIndex
    """
    return {
        "organization_id": auth_context.organization_id
    }

async def get_accessible_resources(
    auth_context: AuthContext,
    resource_type: str
) -> list[str]:
    """
    Get list of resource IDs user has access to.

    Args:
        auth_context: Authentication context
        resource_type: Type of resource (memory, file, agent, chat)

    Returns:
        List of accessible resource IDs
    """
    supabase = get_supabase_client()

    # Build query for accessible resources
    query = supabase.table("resource_permissions").select("resource_id")
    query = query.eq("resource_type", resource_type)
    query = query.eq("organization_id", auth_context.organization_id)

    # Filter by user or organizational unit
    filters = [f"grantee_id.eq.{auth_context.user_id}"]
    if auth_context.organizational_unit_id:
        filters.append(f"grantee_id.eq.{auth_context.organizational_unit_id}")

    query = query.or_(",".join(filters))

    result = query.execute()

    if result.data:
        return [item["resource_id"] for item in result.data]

    return []

# Logging middleware
async def log_access(
    request: Request,
    auth_context: AuthContext,
    action: str,
    resource_type: Optional[str] = None,
    resource_id: Optional[str] = None
):
    """
    Log access attempt to audit table.

    Args:
        request: FastAPI request
        auth_context: Authentication context
        action: Action being performed
        resource_type: Type of resource being accessed
        resource_id: ID of resource being accessed
    """
    try:
        supabase = get_supabase_client()

        audit_entry = {
            "organization_id": auth_context.organization_id,
            "user_id": auth_context.user_id,
            "action": action,
            "resource_type": resource_type,
            "resource_id": resource_id,
            "details": {
                "method": request.method,
                "path": str(request.url.path),
                "user_agent": request.headers.get("user-agent"),
            },
            "ip_address": request.client.host if request.client else None,
        }

        supabase.table("audit_logs").insert(audit_entry).execute()

    except Exception as e:
        logger.error(f"Failed to log audit entry: {str(e)}")
        # Don't fail the request if audit logging fails